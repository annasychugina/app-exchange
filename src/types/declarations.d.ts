declare module 'react-hot-loader';
declare module 'react-redux';
declare module '*.jpg';
declare module '*.jpg?inline';
declare module '*.png';
declare module '*.png?inline';
declare module '*.svg';
declare module '*.svg?inline';
declare module '*.css';
declare module '*.json';

interface RequireImport {
  default: any;
}
