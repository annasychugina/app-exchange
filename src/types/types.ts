import {VariantType} from 'notistack';

export type Currency = 'RUB' | 'USD' | 'EUR' | 'GBP';
export type SymbolCurrency = '$' | '€' | '£' | '₽';

export type FieldInputCallback = (floatValue: number, fieldName: string) => void;

export enum CurrencyBlockType {
  currencyFrom = 'currencyFrom',
  currencyTo = 'currencyTo',
}
export type Rate = {
  [key in Currency]: number;
};

export type RatesState = {
  [key in Currency]?: {
    loading: boolean;
    loaded?: boolean;
    rates: Rate;
  };
};

export type NotificationState = Array<NotificationPayload> | [];
export interface NotificationPayload {
  text: string;
  variant: VariantType;
}

export type CurrentBalanceConfig = {
  [key in string]: number;
};

export interface GlobalState {
  rates: RatesState;
  userBalance: CurrentBalanceConfig;
  notifications: Array<NotificationPayload> | [];
}

export type InputValueState = number | null;

export interface FormCurrencyState {
  currencyFrom: number | null;
  currencyTo: number | null;
}

export interface FormInitialCurrencyState {
  currencyFrom: number | null;
  currencyTo: number | null;
}
