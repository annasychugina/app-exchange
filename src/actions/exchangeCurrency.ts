import {Currency, FormCurrencyState} from '../types/types';

export enum ExchangeActionType {
  EXCHANGE_MONEY = 'EXCHANGE_MONEY',
}
interface ExchangeMoneyPayload {
  currencyFrom: Currency;
  currencyTo: Currency;
  form: FormCurrencyState;
}
interface ExchangeMoneyAction {
  payload: ExchangeMoneyPayload;
  type: ExchangeActionType.EXCHANGE_MONEY;
}

export type ActionType = ExchangeMoneyAction;

export const exchangeCurrency = (payload: ExchangeMoneyPayload): ExchangeMoneyAction => {
  return {
    type: ExchangeActionType.EXCHANGE_MONEY,
    payload,
  };
};
