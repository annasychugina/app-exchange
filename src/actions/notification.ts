import {NotificationPayload} from '../types/types';

export enum NotificationActionType {
  ADD_NOTIFICATION = 'ADD_NOTIFICATION',
  RESET_NOTIFICATIONS = 'RESET_NOTIFICATIONS',
}

export interface NotificationAction {
  payload: Array<NotificationPayload> | [];
  type: NotificationActionType;
}

export const addNotification = (payload: Array<NotificationPayload>): NotificationAction => {
  return {
    type: NotificationActionType.ADD_NOTIFICATION,
    payload,
  };
};

export const resetNotifications = (): NotificationAction => {
  return {
    type: NotificationActionType.RESET_NOTIFICATIONS,
    payload: [],
  };
};
