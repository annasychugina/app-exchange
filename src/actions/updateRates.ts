import {Currency, GlobalState} from '../types/types';
import {Dispatch} from 'redux';
import {getRates, Rate} from '../services/rates';

export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_RESPONSE_SUCCESS = 'RATES_RESPONSE_SUCCESS',
  RATES_RESPONSE_ERROR = 'RATES_RESPONSE_ERROR',
}

interface RatesRequestAction {
  payload: {
    currency: Currency;
  };
  type: ActionType.RATES_REQUEST;
}

interface RatesResponseErrorAction {
  payload: {
    currency: Currency;
  };
  type: ActionType.RATES_RESPONSE_ERROR;
}

interface RatesResponseSuccessAction {
  payload: {
    currency: Currency;
    rates: Rate;
  };
  type: ActionType.RATES_RESPONSE_SUCCESS;
}

const ratesRequest = (currency: Currency) => ({
  payload: {currency},
  type: ActionType.RATES_REQUEST,
});

const ratesResponseSuccess = (currency: Currency, rates: Rate) => ({
  payload: {currency, rates},
  type: ActionType.RATES_RESPONSE_SUCCESS,
});

const ratesResponseError = (currency: Currency) => ({
  payload: {currency},
  type: ActionType.RATES_RESPONSE_ERROR,
});

export const updateRatesForCurrency = (currency: Currency) => (dispatch: Dispatch, getState: () => GlobalState) => {
  const state = getState();
  const rates = state.rates[currency];
  if (rates && rates.loading) {
    return;
  }
  dispatch(ratesRequest(currency));
  return getRates(currency)
    .then(({rates}) => {
      dispatch(ratesResponseSuccess(currency, rates));
    })
    .catch(() => dispatch(ratesResponseError(currency)));
};
export type RateAction = RatesRequestAction | RatesResponseErrorAction | RatesResponseSuccessAction;
