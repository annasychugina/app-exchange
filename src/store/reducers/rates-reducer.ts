import {ActionType, RateAction} from '../../actions/updateRates';
import {RatesState} from '../../types/types';

export const ratesReducer = (state: RatesState = {}, action: RateAction): RatesState => {
  switch (action.type) {
    case ActionType.RATES_REQUEST: {
      const {currency} = action.payload;

      return {
        ...state,
        [currency]: {
          ...state[currency],
          loading: true,
        },
      };
    }

    case ActionType.RATES_RESPONSE_SUCCESS: {
      const {currency, rates} = action.payload;
      return {
        ...state,
        [currency]: {
          ...state[currency],
          rates,
          loaded: true,
          loading: false,
        },
      };
    }

    case ActionType.RATES_RESPONSE_ERROR: {
      const {currency} = action.payload;
      return {
        ...state,
        [currency]: {
          ...state[currency],
          loading: false,
        },
      };
    }

    default:
      return state;
  }
};
