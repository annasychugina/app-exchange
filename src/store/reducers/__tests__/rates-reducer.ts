import {ratesReducer} from '../rates-reducer';
import {ActionType} from '../../../actions/updateRates';
import {Currency} from '../../../types/types';

describe('Rates Reducer', () => {
  it('init state correct', () => {
    expect(ratesReducer(undefined, {} as any)).toEqual({});
  });

  it('RATES_REQUEST action', () => {
    expect(
      ratesReducer(
        {
          ['RUB' as Currency]: {
            loading: false,
          },
        },
        {
          payload: {
            currency: 'RUB',
          },
          type: ActionType.RATES_REQUEST,
        },
      ),
    ).toMatchSnapshot();

    expect(
      ratesReducer(
        {
          ['USD' as Currency]: {
            loading: false,
          },
        },
        {
          payload: {
            currency: 'RUB',
          },
          type: ActionType.RATES_REQUEST,
        },
      ),
    ).toMatchSnapshot();
  });
  it('RATES_RESPONSE_SUCCESS action', () => {
    expect(
      ratesReducer(
        {
          ['USD' as Currency]: {
            loading: true,
          },
        },
        {
          payload: {
            currency: 'USD',
            rates: {
              ['EUR']: 1,
              ['RUB']: 1.5,
            },
          },
          type: ActionType.RATES_RESPONSE_SUCCESS,
        },
      ),
    ).toMatchSnapshot();
  });

  it('RATES_RESPONSE_ERROR Action', () => {
    expect(
      ratesReducer(
        {
          ['RUB' as Currency]: {
            loading: true,
          },
        },
        {
          payload: {currency: 'RUB'},
          type: ActionType.RATES_RESPONSE_ERROR,
        },
      ),
    ).toMatchSnapshot();
  });
});
