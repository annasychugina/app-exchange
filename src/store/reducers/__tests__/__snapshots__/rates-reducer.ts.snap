// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rates Reducer RATES_REQUEST action 1`] = `
Object {
  "RUB": Object {
    "loading": true,
  },
}
`;

exports[`Rates Reducer RATES_REQUEST action 2`] = `
Object {
  "RUB": Object {
    "loading": true,
  },
  "USD": Object {
    "loading": false,
  },
}
`;

exports[`Rates Reducer RATES_RESPONSE_ERROR Action 1`] = `
Object {
  "RUB": Object {
    "loading": false,
  },
}
`;

exports[`Rates Reducer RATES_RESPONSE_SUCCESS action 1`] = `
Object {
  "USD": Object {
    "loaded": true,
    "loading": false,
    "rates": Object {
      "EUR": 1,
      "RUB": 1.5,
    },
  },
}
`;
